framework/src/org/checkerframework/common/fieldtypeinference/FieldTypeInferenceTransfer.java

```diff
@@ -0,0 +1,112 @@
+ * <tt>org.checkerframework.common.fieldtypeinference.FieldTypeInferenceAnnotatedTypeFactory</tt>.
+ *
+ * @author pbsf
+ */
+
+public class FieldTypeInferenceTransfer extends CFTransfer {
+
+    private final Map<Receiver, AnnotatedTypeMirror> assignedPrivateFields;
+    private final FieldTypeInferenceAnnotatedTypeFactory factory;
+
+    public FieldTypeInferenceTransfer(CFAbstractAnalysis<CFValue, CFStore,
+            CFTransfer> analysis) {
+        super(analysis);
+        assert (analysis.getTypeFactory() instanceof FieldTypeInferenceAnnotatedTypeFactory);
+        factory = (FieldTypeInferenceAnnotatedTypeFactory) analysis.getTypeFactory();
+        assignedPrivateFields = factory.getAssignedPrivateFields();
+    }
+
+    @Override
+    public CFStore initialStore(UnderlyingAST underlyingAST,
+            List<LocalVariableNode> parameters) {
+        CFStore initialStore = super.initialStore(underlyingAST, parameters);
+        for (Receiver r : assignedPrivateFields.keySet()) {
+            for (AnnotationMirror am : assignedPrivateFields.get(r).
+                    getAnnotations()) {
+                initialStore.insertValue(r, am);
+            }
+        }
+        return initialStore;
+    }
+
+    @Override
+    public TransferResult<CFValue, CFStore> visitAssignment(AssignmentNode n,
+            TransferInput<CFValue, CFStore> in) {
+        TransferResult<CFValue, CFStore> result = super.visitAssignment(n, in);
+        assert result instanceof RegularTransferResult;
+        Node rhs = n.getExpression();
+        Receiver expr = FlowExpressions.internalReprOf(analysis.getTypeFactory(),
+                n.getTarget());
+
+        // If this is an assignment to a field of 'this' and this field is
+        // private, then populate the map in the ATF.
+        if (!expr.containsUnknown() && expr instanceof FieldAccess &&
+                ((FieldAccess) expr).getField().getModifiers().
+                contains(Modifier.PRIVATE)) {
+            addAssignedField(expr, factory.getAnnotatedType(rhs.getTree()));
+        }
+        return result;
+    }
+
+    /**
+     * Update the ATM mapped by the parameter <tt>r</tt> in the map
+     * <tt>ClassValAnnotatedTypeFactory.analysis.assignedPrivateFields</tt> to be the LUB
+     * of the parameter <tt>atm</tt> and the current ATM mapped by <tt>r</tt>
+     * in this map.
+     * <p>
+     * If <tt>r</tt> is not a key in the map, it is added to the map
+     * mapping to <tt>atm</tt>.
+     * @param r
+     *      A receiver representing a private field.
+     * @param atm
+     *      An ATM assigned to the private field in <tt>r</tt>.
+     */
+    public void addAssignedField(Receiver r, AnnotatedTypeMirror atm) {
+        Map<Receiver, AnnotatedTypeMirror> assignedPrivateFields = factory.getAssignedPrivateFields();
+        if (assignedPrivateFields.containsKey(r)) {
+            Set<? extends AnnotationMirror> lub = analysis.getTypeFactory().
```

Comment:
Hi Paulo,

public Set<? extends AnnotationMirror> leastUpperBounds(AnnotatedTypeMirror
> type1,
> AnnotatedTypeMirror type2, Collection<? extends AnnotationMirror> annos1,Collection<AnnotationMirror>
> annos2)

I am not sure of the purpose of this method and, I think, in the long run
we'll want to remove it.

There is a class, org.checkerframework.framework.util.AnnotatedTypes that
has utility methods for working with AnnotatedTypeMirrors.  One of those is
leastUpperBound.  It will handle typevars correctly (modulo corner cases).

I think I was wrong to suggest AnnotatedTypeMerger because you only want to
refine the primary annotation.  That said, if you use the LUB method I
specify above you may get a type back that is above the declaration of the
field's Java type (see the example below).
List<String> myField;
public void method( @NonNull List<@NonNull String> listOfNonNull, @NonNull
List<@Nullable String> listOfNullable) {
   myField = listOfNonNull;
   myField = listOfNullable;
}

The assignment of nullableList to  myField will issue an error but what I
really am concerned about is that LUBbing the RHS of both assignments would
should actually yield:
@NonNull List<? extends @Nullable String super @NonNull Void>
Though I think with our current LUB code you will actually get a:
@NonNull Object

This, incidentally, has the correct primary annotation for your purposes
but not the correct Java type.  So, if you only want to refine the primary
annotation (and not the Java types/type arguments) you may want to do the
following:

//let fieldDeclaration be the type of receiver's declaration
//let rhsOfAssignment be the type being assigned to the field

AnnotatedTypeMirror previousLub =
typeFactory.getAssignedPrivateFieldATM(receiver);
if (previousLub == null ) {
    AnnotatedTypeMirror copyOfFieldDeclaration =
fieldDeclaration.deepCopy(true);
   copyOfFieldDeclaration.clearAnnotations();
   copyOfFieldDeclaration.addAnnotations(rhsOfAssignment.getAnnotations());

   addAssignedPrivateField(receiver, copyOfFieldDeclaration);
} else {
    AnnotatedTypeMirror lub = AnnotatedTypes.leastUpperBound(previousLub,
rhsOfAssignment);
   previousLub.clearAnnotations();
   previousLub.addAnnotations( lub.getAnnotations());
   //if you use the previous lub you don't have to put it back in the map
(since it already is in the map)
   //also, we know it was already a copy of the declaration so no worries
that it doesn't have the same type as the declaration
}

In the above, I always overwrite the a copy of the field declaration
because the RHS may or may not have the same type.  Now, if we do want to
refine the primary annotation, the Java type, and any component annotations
then you should just put the lub in the map.

Finally, I use clear then add in this case because if the field is a type
variable and the lub has no primary annotations, replaceAnnotation won't
remove the primary annotations that should be removed.

Does this make sense?

Jonathan





On Wed, Jul 8, 2015 at 12:33 PM, Paulo Barros <notifications@github.com>
wrote:

> In
> framework/src/org/checkerframework/common/fieldtypeinference/FieldTypeInferenceTransfer.java
> <https://github.com/typetools/checker-framework/pull/441#discussion_r34188876>
> :
>
> > +     * Update the ATM mapped by the parameter <tt>r</tt> in the map
> > +     * <tt>ClassValAnnotatedTypeFactory.analysis.assignedPrivateFields</tt> to be the LUB
> > +     * of the parameter <tt>atm</tt> and the current ATM mapped by <tt>r</tt>
> > +     * in this map.
> > +     * <p>
> > +     * If <tt>r</tt> is not a key in the map, it is added to the map
> > +     * mapping to <tt>atm</tt>.
> > +     * @param r
> > +     *      A receiver representing a private field.
> > +     * @param atm
> > +     *      An ATM assigned to the private field in <tt>r</tt>.
> > +     */
> > +    public void addAssignedField(Receiver r, AnnotatedTypeMirror atm) {
> > +        Map<Receiver, AnnotatedTypeMirror> assignedPrivateFields = factory.getAssignedPrivateFields();
> > +        if (assignedPrivateFields.containsKey(r)) {
> > +            Set<? extends AnnotationMirror> lub = analysis.getTypeFactory().
>
> This comment makes a lot of sense, and I agree with everything. But unlike
> the other comments, I'm not sure how to fix this one. I understand that we
> must take the lowerBounds and upperBounds of the AnnotatedTypeMirror in
> consideration, as well as the primary annotations, but how to do that
> correctly using the Checker Framework? I don't see a LUB method for
> AnnotatedTypeMirror in the QualifierHierarchy. The closest one I found is
> this:
>
> public Set<? extends AnnotationMirror> leastUpperBounds(AnnotatedTypeMirror type1,
>             AnnotatedTypeMirror type2, Collection<? extends AnnotationMirror> annos1,
>             Collection<AnnotationMirror> annos2)
>
> But as you can see, only primary annotations are returned.
> Is there a reason why a method that calculates the LUB between two
> AnnotatedTypeMirror and returns an AnnotatedTypeMirror does not exist? In
> case not, should we implement it?
>
> Or does a possible solution to that problem has anything to do with the
> AnnotatedTypeMerger you mentioned in another comment? (The comment about
> using the AnnotatedTypeMerger wasn't very clear to me to be honest.)
>
> Thanks a lot for all the feedback!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/typetools/checker-framework/pull/441/files#r34188876>.
>